matchers/adapter/parse.go

```diff
@@ -0,0 +1,142 @@
+
+// MatchersParser is an interface for parsing one or more matchers in the input string.
+type MatchersParser func(s string) (labels.Matchers, error)
+
+// OldMatcherParser uses the old pkg/labels parser to parse the matcher in
+// the input string.
+func OldMatcherParser(l log.Logger) MatcherParser {
+	return func(s string) (*labels.Matcher, error) {
+		level.Debug(l).Log(
+			"msg",
+			"Parsing matcher with old regular expressions parser",
+			"matcher",
+			s,
+		)
+		return labels.ParseMatcher(s)
+	}
+}
+
+// OldMatchersParser uses the old pkg/labels parser to parse zero or more
+// matchers in the string. It returns an error if the input is invalid.
+func OldMatchersParser(l log.Logger) MatchersParser {
+	return func(s string) (labels.Matchers, error) {
+		level.Debug(l).Log(
+			"msg",
+			"Parsing matchers with old regular expressions parser",
+			"matchers",
+			s)
+		return labels.ParseMatchers(s)
+	}
+}
+
+// FallbackMatcherParser uses the new matchers/parse parser to parse
+// zero or more matchers in the string. If this fails it falls back to
+// the old pkg/labels parser and emits a warning log line.
+func FallbackMatcherParser(l log.Logger) MatcherParser {
+	return func(s string) (*labels.Matcher, error) {
+		var (
+			m          *labels.Matcher
+			err        error
+			invalidErr error
+		)
+		level.Debug(l).Log(
+			"msg",
+			"Parsing matcher with new parser",
+			"matcher",
+			s,
+		)
+		m, err = parse.Matcher(s)
+		if err != nil {
+			// The input is not valid in the old pkg/labels parser either,
+			// it cannot be valid input.
+			m, invalidErr = labels.ParseMatcher(s)
+			if invalidErr != nil {
+				return nil, invalidErr
+			}
+			// The input is valid in the old pkg/labels parser, but not the
+			// new matchers/parse parser.
+			level.Warn(l).Log(
+				"msg",
+				"Failed to parse input with matchers/parse, falling back to pkg/labels parser",
+				"matcher",
+				s,
+				"err",
+				err,
+			)
+		}
+		return m, nil
+	}
+}
+
+// FallbackMatchersParser uses the new matchers/parse parser to parse the
+// matcher in the input string. If this fails it falls back to the old
+// pkg/labels parser and emits a warning log line.
+func FallbackMatchersParser(l log.Logger) MatchersParser {
+	return func(s string) (labels.Matchers, error) {
+		var (
+			m          []*labels.Matcher
+			err        error
+			invalidErr error
+		)
+		level.Debug(l).Log(
+			"msg",
+			"Parsing matchers with new parser",
+			"matchers",
+			s,
+		)
+		m, err = parse.Matchers(s)
+		if err != nil {
+			// The input is not valid in the old pkg/labels parser either,
+			// it cannot be valid input.
+			m, invalidErr = labels.ParseMatchers(s)
+			if invalidErr != nil {
+				return nil, invalidErr
+			}
+			// The input is valid in the old pkg/labels parser, but not the
+			// new matchers/parse parser.
+			level.Warn(l).Log(
+				"msg",
+				"Failed to parse input with matchers/parse, falling back to pkg/labels parser",
```

Comment:
@gotjosh I'm thinking if I should make this message more actionable for the end user. Perhaps something like:
```suggestion
				"Alertmanager is moving to a new parser for label matchers, and it looks like this matcher will be incompatible. To fix it make sure the regular expression or value is double-quoted.",
```

It's not true that it **has** to be double-quoted, but double-quoting it should guarantee to fix the issue, so I think it's the best recommendation to make.